crates/matrix-sdk/tests/integration/room/spaces.rs

```diff
@@ -0,0 +1,315 @@
+
+    let room = client.get_room(&DEFAULT_TEST_ROOM_ID).unwrap();
+
+    let spaces: Vec<ParentSpace> = room.parent_spaces().await.unwrap().collect().await;
+    assert_eq!(spaces.len(), 1);
+    assert_let!(Some(ParentSpace::Unverifiable(space_id)) = spaces.get(0));
+    assert_eq!(space_id, *DEFAULT_TEST_SPACE_ID);
+}
+
+#[async_test]
+async fn parent_space_illegitimate() {
+    let (client, server) = logged_in_client().await;
+
+    let sync_token =
+        initial_sync_with_m_space_parent(&client, &server, &test_json::PARENT_SPACE_SYNC).await;
+
+    sync_space(&client, &server, sync_token, vec![]).await;
+
+    let room = client.get_room(&DEFAULT_TEST_ROOM_ID).unwrap();
+
+    let spaces: Vec<ParentSpace> = room.parent_spaces().await.unwrap().collect().await;
+    assert_eq!(spaces.len(), 1);
+    assert_let!(Some(ParentSpace::Illegitimate(space)) = spaces.get(0));
+    assert_eq!(space.room_id(), *DEFAULT_TEST_SPACE_ID);
+}
+
+#[async_test]
+async fn parent_space_reciprocal() {
+    let (client, server) = logged_in_client().await;
+
+    let sync_token =
+        initial_sync_with_m_space_parent(&client, &server, &test_json::PARENT_SPACE_SYNC).await;
+
+    let child_event = json!({
+        "content": {
+            "suggested": true,
+            "via": [
+                "example.org",
+                "other.example.org"
+            ]
+        },
+        "event_id": "$143273582443PhrSn:example.org",
+        "origin_server_ts": 1432735824653_u64,
+        "room_id": *DEFAULT_TEST_SPACE_ID,
+        "sender": "@example2:example.org", // Not equal to sender of m.room.space because
+                                           // equality is only required for power-levels
+        "state_key": *DEFAULT_TEST_ROOM_ID,
+        "type": "m.space.child",
+        "unsigned": {
+            "age": 1234
+        }
+    });
+
+    sync_space(&client, &server, sync_token, vec![child_event]).await;
+
+    let room = client.get_room(&DEFAULT_TEST_ROOM_ID).unwrap();
+
+    let spaces: Vec<ParentSpace> = room.parent_spaces().await.unwrap().collect().await;
+    assert_eq!(spaces.len(), 1);
+    assert_let!(Some(ParentSpace::Reciprocal(space)) = spaces.get(0));
+    assert_eq!(space.room_id(), *DEFAULT_TEST_SPACE_ID);
+}
+
+#[async_test]
+async fn parent_space_redacted_reciprocal() {
+    let (client, server) = logged_in_client().await;
+
+    let sync_token =
+        initial_sync_with_m_space_parent(&client, &server, &test_json::PARENT_SPACE_SYNC).await;
+
+    let child_event = json!({
+        "content": {},  // Redacted -> missing "via" key -> invalidates the relationship
+        "event_id": "$143273582443PhrSn:example.org",
+        "origin_server_ts": 1432735824653_u64,
+        "room_id": *DEFAULT_TEST_SPACE_ID,
+        "sender": "@example2:example.org",
+        "state_key": *DEFAULT_TEST_ROOM_ID,
+        "type": "m.space.child",
+        "unsigned": {
+            "age": 1234,
+            "redacted_by": "$15275047031IXQRi:localhost"
```

Comment:
There is no such thing as `unsigned.redacted_by`. It would have to be `redacted_because` and contain a full redaction event for the overall event to be valid as redacted.

Reply:
I copied it from https://github.com/matrix-org/matrix-rust-sdk/blob/ee0010e831e7f5396b1460578424447b0e4ded24/testing/matrix-sdk-test/src/test_json/sync.rs#L645 and didn't want to waste space with the whole `redacted_because`.

Reply:
That's weird that we already have this faulty construction in the codebase. It's still wrong 😅

Reply:
https://github.com/matrix-org/synapse/issues/8764